pde_superresolution/WENO.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+import numpy as np
+
+_p = (1 / 6) * np.array([
+  [2, -7, 11, 0,  0],
+  [0, -1,  5, 2,  0],
+  [0,  0,  2, 5, -1]
+])
+_B1 = np.sqrt(13/12)*np.array([
+  [1, -2,  1,  0, 0],
+  [0,  1, -2,  1, 0],
+  [0,  0,  1, -2, 1]
+])
+_B2 = 0.5*np.array([
+  [1, -4, 3,  0, 0],
+  [0,  1, 0, -1, 0],
+  [0,  0, 3, -4, 1]
+])
+_d = np.array([[0.1, 0.6, 0.3]]).T
+
+
+class WENO(object):
+  """
+  An implementation of 4th order finite volume WENO, following Chi-Wang Shu, "Essentially Non-Oscillatory and Weighted
+  Essentially Non-Oscillatory Schemes for Hyperbolic Conservation Laws", NASA/CR-97-206253, Nov 1997,
+  available at https://www3.nd.edu/~zxu2/acms60790S13/Shu-WENO-notes.pdf
+
+  In their notation, this implementation uses k=3.
+  """
+
+  def __init__(self, dx, flux=lambda x: -0.5 * x ** 2, dflux=lambda x: -x):
+    """Constructor.
+        Args:
+          dx: grid spacing
+          flux, dflux: callables that return the physical flux f(w) and its
+          derivative f'(w)
+    """
+    self.flux = flux
+    self.dflux = dflux
+    self.eps = 1e-6
+    self.dx = dx
+
+  def flux_divergence(self, w: np.ndarray) -> np.ndarray:
+    """
+    returns the WENO approximation of the divergence of the flux
+    """
+    f = self.flux(w)
+
+    # Lax-Friedrichs flux splitting
+    a = np.max(np.abs(self.dflux(w)))
```

Comment:
Should this be a global or local maximum? I'm testing `ndimage.filters.maximum_filter(abs(self.dflux(w)), size=5, mode='wrap')` here as well.

Reply:
I was following this: https://github.com/wme7/WENO/blob/b39be53fc61f4f2743a1a9a56c32e5c6cf23f4f6/Non-lineard1/WENO5resAdv1d.m#L54